### PR TITLE
fix: python: Differentiate methods from functions in Generic

### PR DESCRIPTION
**Pull request by akuhlens**
This is used in semgrep/semgrep-proprietary#1214 to properly generate the taint parameter environment for Python instance methods. Unlike Java/C#/etc, in Python the receiver object of an instance method is declared explicit in the method's definition.

## Test Plan
New tests working in PR semgrep/semgrep-proprietary#1214

### CODE REVIEW

File: src/tainting/Taint.ml [hunk 1]
```diff
@@ -374,6 +375,13 @@ end)
 
 type signature = Findings.t
 
+let show_signature s =
```
Comment (IagoAbal) [inline src/tainting/Taint.ml]: I should probably use the `debug_` prefix, but until now I was using `_show_` to mean the same.

File: src/tainting/Dataflow_tainting.ml [hunk 2]
```diff
@@ -1337,6 +1349,7 @@ let check_function_signature env fun_exp args args_taints =
 
          So we will isolate this as a specific step to be applied as necessary.
       *)
+      logger#trace "check_function_sig fun_sid: %s" (Taint.show_signature fun_sig);
```
Comment (IagoAbal) [inline src/tainting/Dataflow_tainting.ml]: If you're using these for debugging better use `logger#flash` IMO, and we have rules that will make sure no code is merged containing `logger#flash`.

File: src/analyzing/AST_to_IL.ml [hunk 3]
```diff
@@ -1804,8 +1804,24 @@ and function_definition env fdef =
...
+  in
+  logger#trace "is_method: %s" (if is_method then "true" else "false");
   let env = { (empty_env lang) with ctx = ctx ||| empty_ctx } in
-  let params = parameters env def.G.fparams in
+  (* TODO add the self or class parameter to the env *)
+  logger#trace "def parameters: %s" (G.show_parameters def.G.fparams);
+  let params =
+    let ps = if is_method then
+        let (lb, ps', rb) = def.G.fparams in
+        (lb, (try List.tl ps'  with Failure _ -> ps'), rb)
```
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: Better to `match` on the list IMO than to `try ... with Failure ...`.

File: src/tainting/Taint.mli [hunk 4]
```diff
@@ -231,3 +231,4 @@ type signature = Findings.t
  *)
 
 val _show_finding : finding -> string
+val _show_signature : signature -> string
```
Comment (nmote) [inline src/tainting/Taint.mli:234]: Typically we use leading underscores only for symbols that are not used or exposed publicly. I get that you're just following what's already here in this case, but do you know why this was done here?

**Review by nmote — APPROVED** [review]

File: src/tainting/Taint.mli (hunk 4 above)
Comment (akuhlens) [reply src/tainting/Taint.mli:234]: These are just exposed for debugging according to a comment on the previous comment. I will follow this up with renaming all of these to be `debug_show_*` per that comment.

File: languages/python/generic/Python_to_generic.ml [hunk 5]
```diff
@@ -560,6 +560,11 @@ and fieldstmt x =
 and stmt_aux env x =
   match x with
   | FunctionDef (t, v1, v2, v3, v4, v5) ->
+      let fkind =
```
Comment (IagoAbal) [inline languages/python/generic/Python_to_generic.ml:563]: Given these changes I find the PR title a bit unrelated (and way too long... you can elaborate in the PR description).

Comment (IagoAbal) [reply languages/python/generic/Python_to_generic.ml:563]: This PR basically distinguishes functions versus methods in Python-to-Generic translation. I think the PR title should reflect just that, then you can always say what you want to do with that, or link to another related PR or ticket.

**Review by IagoAbal — APPROVED** [review]
We should put a bit more time into good PR descriptions, this change was tiny, but a larger change with a similar PR description would be harder to review, and harder to understand in the future if the PR happens to introduce a bug that is only found long time after.

**Pull request merged**